---
## Hausaufgabe 1 - Grundlagen Python & Objektorientierung
---

### Achtung: Überprüfen Sie vor Abgabe der Hausaufgabe, ob das Notebook richtig gespeichert wurde. 
Die Speicherung von Notebooks funktioniert über das "Disketten"-Symbol im Notebook oder über die Shortcuts Strg + S (Win) und CMD + S (MAC).

---
## Bewertung
---

#### Erreichbare Punkte: 25
---
#### Erreichte Punkte:   -
---
---
Aufgabe 1:<br>
<br>
Aufgabe 2:<br>
<br>
Aufgabe 3:<br>

### Aufgabe 1 - Listen & Sortierungen (6 Punkte)

Ihnen ist die folgende Liste gegeben.


In [1]:
thisList = ['Otto', 'Level', 'Leseesel', [1, 2, 3], 'Kajak', 'Mueller', 'Retter', 'Rotor', [5, 6, 7], 'Effe', 'Meier']

#### a.) 
#### (2 Punkte)

Erzeugen Sie eine zweite Liste *thisList2*, die die Einträge von *thisList* übernimmt. Jedoch sollen alle Elemente in der neuen Liste umgekehrt sein.
Informieren Sie sich dazu über die Möglichkeiten des Index-Operators und nutzen Sie diesen, um die gewünschte Ausgabe zu erzeugen.

In [2]:
### Ergänzen Sie hier Ihren Code, um die gewünschte Ausgabe zu erzielen. ###
thisList2 = []
for item in thisList:
    length = len(item) - 1
    if type(item) == str:
        item2 = ''
        while length >= 0:
            item2 += item[length]
            length -= 1
    else:
        item2 = []
        while length >= 0:
            item2.append(item[length])
            length -= 1
    thisList2.append(item2)




### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen ###

print(thisList2)

['ottO', 'leveL', 'leseeseL', [3, 2, 1], 'kajaK', 'relleuM', 'retteR', 'rotoR', [7, 6, 5], 'effE', 'reieM']


Erwarteter Output:
    
    '['ottO', 'leveL', 'leseeseL', [3, 2, 1], 'kajaK', 'relleuM', 'retteR', 'rotoR', [7, 6, 5], 'effE', 'reieM']'





#### b.)
#### (1 Punkte)

Erzeugen Sie eine dritte Liste *thisList3*, die nur diejenigen Einträge aus *thisList* übernimmt, die den Datentype `str` aufweisen.

In [3]:
### Ergänzen Sie hier Ihren Code, um die gewünschte Ausgabe zu erzielen. ###
thisList3 = []

for item in thisList:
    if type(item) == str:
        thisList3.append(item)

### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen ###

print(thisList3)

['Otto', 'Level', 'Leseesel', 'Kajak', 'Mueller', 'Retter', 'Rotor', 'Effe', 'Meier']


Erwarteter Output:
    
    '['Otto', 'Level', 'Leseesel', 'Kajak', 'Mueller', 'Retter', 'Rotor', 'Effe', 'Meier']'

#### c.)
#### (2 Punkte)

In der Liste *evaluations* finden Sie die bereits bekannten Einträge aus *thisList3* zusammen mit weiteren Werten in Tupeln. Sortieren Sie die Einträge von *evaluations* absteigend anhand der **Summe** der Zahleneinträge eines jeden Tupels mithilfe einer `lambda`-Funktion. Geben Sie anschließend das Tupel mit der höchsten Summe aus.

In [4]:
#import random
evaluations = [('Otto', 3.779, 1.238, 0.49), ('Level', 3.961, 5.725, 0.233), ('Leseesel', 3.935, 1.482, 2.41), ('Kajak', 1.989, 2.66, 0.656), ('Mueller', 2.024, 1.427, 3.013), ('Retter', 3.297, 2.36, 3.179), ('Rotor', 1.733, 4.218, 4.972), ('Effe', 3.311, 3.197, 3.991), ('Meier', 5.956, 2.554, 4.622)]

### Bitte ergänzen Sie hier Ihren Code ###

eva_desc = sorted(evaluations, key = lambda x: sum(x[1: ]), reverse = True)
print(eva_desc[0])

('Meier', 5.956, 2.554, 4.622)


Erwarteter Output:

    ('Meier', 5.956, 2.554, 4.622)

#### d.)
#### (1 Punkte)

Oh nein! Leider wurde die Liste leicht beschädigt, sodass nun zusätzliche Listen innerhalb der eigentlichen Listen entstanden sind. Bitte führen Sie die Liste *thisList3Corrupted* mit einem **Python-Ausdruck** wieder in den Ursprungszustand zurück.

In [5]:
thisList3Corrupted = ['Otto', ['Level', 'Leseesel', 'Kajak'], ['Mueller'], 'Retter', 'Rotor', ['Effe', 'Meier']]

### Bitte ergänzen Sie hier Ihren Code ###
thisList3Repaired = []
for item in thisList3Corrupted:
    if type(item) == str:
        thisList3Repaired.append(item)
    else: 
        for element in item:
            thisList3Repaired.append(element)

### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen ###

print(thisList3Repaired)

['Otto', 'Level', 'Leseesel', 'Kajak', 'Mueller', 'Retter', 'Rotor', 'Effe', 'Meier']


Erwarteter Output:

    ['Otto', 'Level', 'Leseesel', 'Kajak', 'Mueller', 'Retter', 'Rotor', 'Effe', 'Meier']

### Aufgabe 2 - Einfache Funktionen (7 Punkte)

#### a.)
#### (4 Punkte)

Sie haben gemerkt, dass manche Einträge der Liste *thisList* eine besondere Eigenschaft besitzen. Es handelt sich bei einigen Einträgen um Palindrome, die rückwärts gelesen genau denselben Text ergeben, wie in der eigentlichen Leserichtung.

Um nun möglichst schnell herauszufinden, welche Einträge einer Liste Palindrome darstellen, schreiben Sie bitte eine Funktion `CheckForPalindroms`, der eine **Liste** übergeben wird. Als Rückgabewert soll die Funktion anschließend in einer Liste die Indices der Elemente zurückgeben, die als Palindrom identifiziert wurden.

**Tipp:**
<details>

Achten Sie darauf, dass gilt: 'a' != 'A'. Finden Sie also zunächst eine Funktion, um die Strings zu formatieren.
</details>

In [6]:
thisList = ['Otto', 'Level', 'Leseesel', [1, 2, 3], 'Kajak', 'Mueller', 'Retter', 'Rotor', [5, 6, 7], 'Effe', 'Meier']

### Bitte ergänzen Sie hier Ihren Programmcode ###
def CheckForPalindroms(list):
    ind_list = []
    for i in range(len(list)):
        if type(list[i]) == str:
            low=list[i].lower()
            
            if low[::-1] == low:
                ind_list.append(i)
    return ind_list

        


### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen ###

CheckForPalindroms(thisList)

[0, 1, 2, 4, 6, 7, 9]

Erwarteter Output:

    
    [0, 1, 2, 4, 6, 7, 9]








#### b.)
#### (3 Punkte)

Schreiben Sie eine Funktion ``splitNumber``, die eine ihr übergebene Gleitkommazahl in deren ganzzahligen Teil und den Dezimal-Teil aufteilt und beides zusammen als Tupel ausgibt. Nutzen Sie dafür **ausschließlich die Ihnen bekannten Operatoren** und **keine built-in-Funktionen**.

Für den Fall, dass keine Zahl übergeben wird, soll eine `Exception` als `TypeError` ausgegeben werden. Informieren Sie sich dazu zum Konstrukt *try-except* und wenden dieses an.

In [7]:
### Bitte ergänzen Sie hier Ihren Programmcode ###
def splitNumber(num):
    try:
        num_ganz = int(num)
        num_dez = num-num_ganz
        return (num_ganz, num_dez)


    except ValueError as e:
        print(e)
        print("TypeError: Input has wrong type!")

### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen ###

print(splitNumber(3.56))
print(splitNumber("Hallo"))

(3, 0.56)
invalid literal for int() with base 10: 'Hallo'
TypeError: Input has wrong type!
None


Erwarteter Output:

    (3, 0.56)
    
    ...(Hier steht eine Fehlerbeschreibung)...

    TypeError: Input has wrong type!

### Aufgabe 3 - Klassendefinition & Weitere Funktionen (12 Punkte)

#### a.) Die Rechteck-Klasse
#### (8 Punkte)

Erstellen Sie eine Klasse *Rechteck*, die als Parameter eine Id *id*, eine Breite *width* und eine Höhe *height* entgegennimmt. Die Methode `DetermineArea()` soll die Fläche des Rechtecks berechnen. Diese Methode soll direkt in der `__init__()`-Methode aufgerufen werden. Zusätzlich soll die `__str__()`-Methode für eine entsprechende Ausgabe sorgen.

Anschließend definieren Sie eine Funktion `Load` (diese ist nicht Teil der Klasse *Rechteck*), welche die Datei `Rectangulars.json` einliest, aus den enthaltenen Informationen Objekte der Klasse *Rechteck* erzeugt und alle Objekte in einer Liste ausgibt.

In [8]:
import json
### Bitte ergänzen Sie hier Ihren Programmcode ###
class Rechteck():

    def __init__(self, id, width, height):
        self.id = id
        self.width = width
        self.height = height
        self.area = self.DetermineArea()

    def DetermineArea(self):
        return self.width * self.height

    def __str__(self):
        return f'Recteck {self.id} mit Breite {self.width}, Höhe {self.height}, Fläche {self.area}'
    
def Load(name):
    rec_list = []
    with open(name) as file:
        data = json.load(file)
        for id in data.keys():
            rec = Rechteck(id, data[id]["width"], data[id]["height"])
            rec_list.append(rec)

    return rec_list



### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen ###

rectangulars = Load('Rectangulars.json')

for rec in rectangulars:
    print(rec)


Recteck 0 mit Breite 6, Höhe 6, Fläche 36
Recteck 1 mit Breite 7, Höhe 10, Fläche 70
Recteck 2 mit Breite 9, Höhe 6, Fläche 54
Recteck 3 mit Breite 7, Höhe 10, Fläche 70
Recteck 4 mit Breite 5, Höhe 11, Fläche 55
Recteck 5 mit Breite 10, Höhe 10, Fläche 100
Recteck 6 mit Breite 10, Höhe 13, Fläche 130
Recteck 7 mit Breite 8, Höhe 6, Fläche 48
Recteck 8 mit Breite 8, Höhe 7, Fläche 56
Recteck 9 mit Breite 9, Höhe 10, Fläche 90


Erwarteter Output:

    Rechteck 0 mit Breite 6, Höhe 6, Fläche 36
    Rechteck 1 mit Breite 7, Höhe 10, Fläche 70
    Rechteck 2 mit Breite 9, Höhe 6, Fläche 54
    Rechteck 3 mit Breite 7, Höhe 10, Fläche 70
    Rechteck 4 mit Breite 5, Höhe 11, Fläche 55
    Rechteck 5 mit Breite 10, Höhe 10, Fläche 100
    Rechteck 6 mit Breite 10, Höhe 13, Fläche 130
    Rechteck 7 mit Breite 8, Höhe 6, Fläche 48
    Rechteck 8 mit Breite 8, Höhe 7, Fläche 56
    Rechteck 9 mit Breite 9, Höhe 10, Fläche 90

#### b.) Eine letzte Nachricht
#### (4 Punkte)

Ihnen wurde eine mysteriöse Nachricht zugeschickt. Zusätzlich zu dieser Nachricht standen auf einem weiteren Zettel nur die Worte "Veni, vidi, vici." und der Hinweis "Schlüssel ist die 20".

Schreiben Sie eine Funktion `Decrypt`, die die verschlüsselte Nachricht decodieren kann. Achten Sie dabei auf Leerzeichen sowie Sonderzeichen.

**Tipp 1:**
<details>

Informieren Sie sich über die Cäsar-Verschlüsselung und deren Funktionsweise!
</details>

**Tipp 2:**
<details>

Informieren Sie sich über die Funktion ``ord(x)`` und achten Sie auf Groß- und Kleinschreibung. Des Weiteren ist auch die Funktion ``.isalpha()`` hilfreich. Letztere lässt sich auf Strings anwenden. 
</details>

In [9]:
verschluesselteNachricht = "Ylmny Buomuozauvy aymwbuzzn! Scjjcy :)"
schluessel = 20

### Bitte ergänzen Sie hier Ihren Programmcode ###
def Decrypt(text, key):
    klartext = ''
    for sign in text:
        if sign >= 'A' and sign <= "Z":
            klartext += chr(ord("A")+((ord(sign)-ord("A"))-key) % 26)

        elif sign >= 'a' and sign <= "z":
            klartext += chr(ord("a")+((ord(sign)-ord("a"))-key) % 26)

        else:
            klartext += sign
    return klartext
            



### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen ###

Decrypt(verschluesselteNachricht, schluessel)

'Erste Hausaufgabe geschafft! Yippie :)'